data/src/main/scala/cats/data/Or.scala

```diff
@@ -1,498 +1,152 @@
- *
- * Often used to represent the failure case of a result
- */
-final case class LOr[+A](a: A) extends (A Or Nothing)
+  def compare[AA >: A, BB >: B](that: AA Or BB)(implicit AA: Order[AA], BB: Order[BB]): Int = fold(
+    a => that.fold(AA.compare(a, _), _ => -1),
+    b => that.fold(_ => 1, BB.compare(b, _))
+  )
 
-/** A right disjunction
- *
- * Often used to represent the success case of a result
- */
-final case class ROr[+B](b: B) extends (Nothing Or B)
+  def partialCompare[AA >: A, BB >: B](that: AA Or BB)(implicit AA: PartialOrder[AA], BB: PartialOrder[BB]): Double = fold(
+    a => that.fold(AA.partialCompare(a, _), _ => -1),
+    b => that.fold(_ => 1, BB.partialCompare(b, _))
+  )
 
-object Or extends OrInstances with OrFunctions {
+  def ===[AA >: A, BB >: B](that: AA Or BB)(implicit AA: Eq[AA], BB: Eq[BB]): Boolean = fold(
+    a => that.fold(AA.eqv(a, _), _ => false),
+    b => that.fold(_ => false, BB.eqv(b, _))
+  )
 
-  /** Spin in tail-position on the right value of the given disjunction. */
-  @annotation.tailrec
-  final def loopRight[A, B, X](d: A Or B, left: A => X, right: B => X Or (A Or B)): X =
-    d match {
-      case LOr(a) => left(a)
-      case ROr(b) => right(b) match {
-        case LOr(x) => x
-        case ROr(q) => loopRight(q, left, right)
-      }
-    }
-
-  /** Spin in tail-position on the left value of the given disjunction. */
-  @annotation.tailrec
-  final def loopLeft[A, B, X](d: A Or B, left: A => X Or (A Or B), right: B => X): X =
-    d match {
-      case LOr(a) => left(a) match {
-        case LOr(x) => x
-        case ROr(q) => loopLeft(q, left, right)
-      }
-      case ROr(b) => right(b)
-    }
+  def traverse[F[_], AA >: A, D](f: B => F[D])(implicit F: Applicative[F]): F[AA Or D] = this match {
+    case RightOr(b) => F.map(f(b))(Or.right _)
+    case left @ LeftOr(_) => F.pure(left)
+  }
+
+  def foldLeft[C](c: C)(f: (C, B) => C): C = fold(_ => c, f(c, _))
+
+  def foldRight[C](c: C)(f: (B, C) => C): C = fold(_ => c, f(_, c))
 
+  def foldRight[C](c: Lazy[C])(f: (B, Lazy[C]) => C): Lazy[C] = fold(_ => c, b => Lazy(f(b, c)))
+
+  def merge[AA >: A](implicit ev: B <:< AA): AA = fold(identity, ev.apply)
 }
 
-sealed abstract class OrInstances extends OrInstances0 {
-  implicit def OrOrder[A: Order, B: Order]: Order[A Or B] =
-    new Order[A Or B] {
-      def compare(a1: A Or B, a2: A Or B) =
-        a1 compare a2
-      override def eqv(a1: A Or B, a2: A Or B) =
-        a1 === a2
-    }
-
-  implicit def OrMonoid[A: Semigroup, B: Monoid]: Monoid[A Or B] =
-    new Monoid[A Or B] {
-      def combine(a1: A Or B, a2: A Or B) =
-        a1 +++ a2
-      def empty =
-        ROr(Monoid[B].empty)
-    }
-
-  implicit def OrShow[A, B](implicit showA: Show[A], showB: Show[B]): Show[A Or B] =
-    Show.show[A Or B](_.fold(a => s"LOr(${showA.show(a)})", b => s"ROr(${showB.show(b)})"))
+object Or extends OrInstances with OrFunctions {
+  final case class LeftOr[+A](a: A) extends (A Or Nothing)
```

Comment:
do we need the `Or` suffix on these names given they are namespaced by the `Or` object already?